firewood/src/merkle.rs

```diff
@@ -51,27 +53,27 @@ fn nibbles_formatter<X: IntoIterator<Item = u8>>(nib: X) -> String {
 macro_rules! write_attributes {
     ($writer:ident, $node:expr, $value:expr) => {
         if !$node.partial_path.0.is_empty() {
-            write!(
-                $writer,
-                " pp={}",
-                nibbles_formatter($node.partial_path.0.clone())
-            )
-            .map_err(|e| FileIoError::from_generic_no_file(e, "write attributes"))?;
+            let pp_str = format!(" pp={}", nibbles_formatter($node.partial_path.0.clone()));
+            $writer
+                .write_all(pp_str.as_bytes())
+                .map_err(|e| FileIoError::from_generic_no_file(e, "write attributes"))?;
         }
         if !$value.is_empty() {
             match std::str::from_utf8($value) {
                 Ok(string) if string.chars().all(char::is_alphanumeric) => {
-                    write!($writer, " val={}", string)
+                    let val_str = format!(" val={}", string);
+                    $writer
+                        .write_all(val_str.as_bytes())
                         .map_err(|e| FileIoError::from_generic_no_file(e, "write attributes"))?;
                 }
                 _ => {
                     let hex = hex::encode($value);
+                    let val_str = format!(" val={:.6}", hex);
+                    $writer
+                        .write_all(val_str.as_bytes())
+                        .map_err(|e| FileIoError::from_generic_no_file(e, "write attributes"))?;
                     if hex.len() > 6 {
-                        write!($writer, " val={:.6}...", hex).map_err(|e| {
-                            FileIoError::from_generic_no_file(e, "write attributes")
-                        })?;
-                    } else {
-                        write!($writer, " val={}", hex).map_err(|e| {
+                        $writer.write_all(b"...").map_err(|e| {
```

Comment:
Same here. Why are we allocating a heap only to copy it to a buffer and then drop it?

Reply:
Agreed, this all needs a makeover. I'll update the other task I created to include these comments as well.

Reply:
moving it to an issue is unnecessary because this PR is the change that's making it worse. It should be fixed here.